src/stk/databases/mongo_db/constructed_molecule.py

```diff
@@ -689,38 +691,78 @@ def _get_molecule_keys(self, entry):
                 yield key, value
 
     def get_all(self):
-        for entry in self._constructed_molecules.find():
-            # Do 'or' query over all key value pairs.
-            query = {'$or': [
-                {key: value}
-                for key, value in self._get_molecule_keys(entry)
-            ]}
-
-            molecule_json = self._molecules.find_one(query)
-            if molecule_json is None:
-                raise KeyError(
-                    'No molecule found in the database associated '
-                    f'with a position matrix with query: {query}. '
-                    'This suggests your database is corrupted.'
-                )
-
-            position_matrix = self._position_matrices.find_one(query)
-            if position_matrix is None:
-                raise KeyError(
-                    'No position matrix found in the database '
-                    'associated with a position matrix with query: '
-                    f'{query}. This suggests your database is '
-                    'corrupted.'
-                )
-
-            yield self._dejsonizer.from_json(
-                json={
-                    'molecule': molecule_json,
-                    'constructedMolecule': entry,
-                    'matrix': position_matrix,
-                    'buildingBlocks': tuple(map(
-                        self._get_building_block,
-                        entry['BB'],
-                    ))
+        # Get all potential indices.
+        indices = chain(
+            self._position_matrices.index_information().values(),
+            self._molecules.index_information().values(),
+            self._constructed_molecules.index_information().values(),
+        )
+        keys = dedupe(index['key'][0][0] for index in indices)
+
+        # Iterate over potential keys, and aggregate matching position
+        # matrices with molecules and constructedmolecules.
+        for key in keys:
```

Comment:
if we go through all the keys, we can end up with yield each molecule twice, which is different to the current behaviour and potentially undesirable. For example , if we used both the inchikey and smiles, we would yield each molecule twice, once for inchikey and once for smiles. The simple solution is to have set which keeps track of yielded molecules, but this is not a reasonable solution because it takes O(N) memory, meaning using `get_all` would become impossible for sufficiently large databases.

I think what needs to happen is that we need to programatically create the query passed to the `aggregate` method. Meaning that `$lookup` dictionary is inserted into the query for each key. Similarly a `$match` query will be added too. For example 

```python

# need to make a tuple cos keys is gonna be reused
keys = tuple(dedupe(...))
query = [
    {
        '$match': {
            '$expr': {
                  '$or': [
                           {key: {'$exists': True}}
                           for key  in keys
                   ], 
             }, 
        }
    },
]
query.extend(
    {
        '$lookup': {  
              'from': self._molecules.name,
              'localField': key,
              'foreignField': key,
              'as':f 'mol_{key}',
        },
    }
    for key in keys
)
# do lookups for posiation matrix and stuff in the same way
query.append(
    {   
         '$match': {
                 '$expr': {
                          '$or': [
                                      {
                                         '$and': [
                                               {
                                                  '$gt': [
                                                      {'$size': f'$mol_{key}'},
                                                      0
                                                  ],
                                              },
                                              {
                                                  '$gt': [
                                                      {'$size': f'$posmat_{key}'},
                                                      0
                                                  ],
                                              },                                    
                                          ],
                                      }
                                     for key in keys
                            ]
                  },
         },
    }
)

```
(disclaimer - I'm not sure if I've botched this code- my main worry is the `$and` nested in the `$or` - it is for illustrative purposes only)

I think a test case making sure that we get the correct behaviour for a database that has the following entries:
* some molecules with 2 different molecule keys,
* some molecules with only one of the molecule keys
* The database has an index made for both of the molecule keys